RESOURCE_ACCESS_CONTROL_FOR_PLUGINS.md

```diff
@@ -0,0 +1,691 @@
+    "resource_id": "my-resource",
+    "created_by": {
+      "user": "you"
+    },
+    "share_with": {
+      "default": {
+        "users": ["shared-user-name"],
+        "roles": ["shared-roles"],
+        "backend_roles": ["shared-backend-roles"]
+      }
+    }
+  }
+}
+```
+
+Note: `default` is a place-holder action-group that will eventually be replaced with actual action-groups in the future.
+
+#### **Response Fields:**
+| Field          | Type    | Description |
+|---------------|---------|-------------|
+| `sharing_info` | Object  | Contains information about how the resource is shared. |
+| `source_idx`   | String  | The OpenSearch index containing the resource. |
+| `resource_id`  | String  | The unique identifier of the resource being shared. |
+| `created_by`   | Object  | Information about the user who created the sharing entry. |
+| `share_with`   | Object  | Defines users, roles, and backend roles with access to the resource. |
+
+
+### **3. Revoke Access**
+- **Endpoint:**
+  ```
+  POST /_plugins/_security/resources/revoke
+  ```
+- **Description:**
+  Revokes access to a resource for specific users, roles, or backend roles.
+
+#### **Request Body:**
+```json
+{
+  "resource_id": "my-resource",
+  "resource_index": "resource-index",
+  "entities_to_revoke": {
+    "roles": ["shared-roles"]
+  }
+}
+```
+
+#### **Request Fields:**
+| Field            | Type    | Description |
+|-----------------|---------|-------------|
+| `resource_id`   | String  | The unique identifier of the resource whose access is being revoked. |
+| `resource_index`| String  | The OpenSearch index where the resource is stored. |
+| `entities_to_revoke` | Object | Specifies which **users, roles, or backend roles** should have their access removed. |
+| `roles`        | Array   | List of roles to revoke access from. |
+
+#### **Response:**
+Returns the updated **resource sharing state** after revocation.
+```json
+{
+  "sharing_info": {
+    "source_idx": "resource-index",
+    "resource_id": "my-resource",
+    "created_by": {
+      "user": "admin"
+    },
+    "share_with": {
+      "default": {
+        "users": ["shared-user-name"],
+        "backend_roles": ["shared-backend-roles"]
+      }
+    }
+  }
+}
+```
+
+Note: `default` is a place-holder action-group that will eventually be replaced with actual action-groups in the future.
+
+#### **Response Fields:**
+| Field          | Type    | Description |
+|---------------|---------|-------------|
+| `sharing_info` | Object  | Contains information about the updated resource sharing state. |
+| `source_idx`   | String  | The OpenSearch index containing the resource. |
+| `resource_id`  | String  | The unique identifier of the resource. |
+| `created_by`   | Object  | Information about the user who created the sharing entry. |
+| `share_with`   | Object  | Defines users, roles, and backend roles that still have access to the resource. |
+
+
+### **4. List Accessible Resources**
```

Comment:
I think this is better left for each resource plugin to define the endpoints for their plugin. What's the advantage of having a REST API in the security plugin for this?

Reply:
This simply allows listing all resources you have access to. Like say you open a photo app. You should be able to see all photos shared with you regardless of your permission level. However you should not be able to see the photos that have not been shared with you at all.

Reply:
But doesn't the photo APP define the API it needs? What's security's role. My understanding is that the ResourceSharingClient provided to the plugin can be used by the plugin for filtering resources visible to the user.

Reply:
It does but how does it fetch all accessible photos. We don't have a filtering mechanism (aside from DLS, but we benched that approach) for plugins to call a list api and they be able to list all photos. I've implemented this usage in SamplResourcePlugin.

Reply:
> It does but how does it fetch all accessible photos. 

Cant we expose a method in the client that yields the IDs of all accessible photos?

In any case, a photo app also needs to provide an endpoint to list all resources, if it wants to remain usable in case the security plugin is disabled (and therefore the `GET /_plugins/_security/resources/list/{resource_index}` endpoint does not exist).

Any external client will likely use the photo app specific endpoint, as it also would want to support the "security disabled" case.

Reply:
> Cant we expose a method in the client that yields the IDs of all accessible photos?

We do that and a step further by listing the photos itself. If the consensus is to only return IDs I will go ahead and make that change.